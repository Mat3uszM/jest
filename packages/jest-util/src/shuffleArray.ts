import Prando from 'prando';

export const rngBuilder = (seed: number) => new Prando(seed);

// Fisher-Yates shuffle
// This is performed in-place
export default function shuffleArray<T>(
  array: Array<T>,
  random: () => number = Math.random,
): Array<T> {
  const length = array == null ? 0 : array.length;
  if (!length) {
    return [];
  }
  let index = -1;
  const lastIndex = length - 1;
  while (++index < length) {
    const rand = index + Math.floor(random() * (lastIndex - index + 1));
    const value = array[index];
    array[index] = array[rand];
    array[rand] = value;
  }
  return array;
}
