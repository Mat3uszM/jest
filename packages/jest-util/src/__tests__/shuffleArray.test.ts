import shuffleArray, {rngBuilder} from '../shuffleArray';

describe(rngBuilder, () => {
  // Breaking these orders would be a breaking change
  // Some people will be using seeds relying on a particular order
  test.each([1, 2, 4, 8, 16])('creates a randomiser given seed %s', seed => {
    const rng = rngBuilder(seed);
    const results = Array(10)
      .fill(0)
      .map(() => rng.next());
    expect(results).toMatchSnapshot();
  });
});

describe(shuffleArray, () => {
  it('empty array is shuffled', () => {
    const shuffled = shuffleArray([]);
    expect(shuffled).toEqual([]);
  });

  // Breaking these orders would be a breaking change
  // Some people will be using seeds relying on a particular order
  const seed = 123;
  test.each([[['a']], [['a', 'b']], [['a', 'b', 'c']], [['a', 'b', 'c', 'd']]])(
    'shuffles list %p',
    l => {
      const rng = rngBuilder(seed);
      expect(shuffleArray(l, () => rng.next())).toMatchSnapshot();
    },
  );
});
