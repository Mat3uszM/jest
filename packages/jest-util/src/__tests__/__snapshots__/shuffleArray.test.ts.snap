// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rngBuilder creates a randomiser given seed 1 1`] = `
Array [
  0.500062950304724,
  0.5157397988987481,
  0.9226656041393675,
  0.3155057489209589,
  0.1551597700815554,
  0.9486529105223372,
  0.3748853226133821,
  0.3723766031610725,
  0.48958173894546503,
  0.6052440010489067,
]
`;

exports[`rngBuilder creates a randomiser given seed 2 1`] = `
Array [
  0.5001259004930327,
  0.5312583452396231,
  0.6624884909630028,
  0.8890491144007652,
  0.38940668720505356,
  0.6909172021995571,
  0.6007535305341597,
  0.06686430426008634,
  0.881952763041936,
  0.5476082643837688,
]
`;

exports[`rngBuilder creates a randomiser given seed 4 1`] = `
Array [
  0.5002518008696502,
  0.5625166903628308,
  0.8249769818095902,
  0.27809822845228443,
  0.2790498855242156,
  0.0634860098509784,
  0.4434312750686499,
  0.05855132221676207,
  0.5894388301739094,
  0.24544880754441228,
]
`;

exports[`rngBuilder creates a randomiser given seed 8 1`] = `
Array [
  0.500503601622885,
  0.6250333806092463,
  0.14995397095334576,
  0.058307676123992465,
  0.47901072620391166,
  0.9113757456446475,
  0.5750688020081886,
  0.8338097936552507,
  0.8361590078184752,
  0.8376564038539437,
]
`;

exports[`rngBuilder creates a randomiser given seed 16 1`] = `
Array [
  0.5010072108127659,
  0.7520182327721311,
  0.2882353561670136,
  0.3866406226033905,
  0.8962605120372634,
  0.19173304927342874,
  0.08497578978654365,
  0.3369975006992457,
  0.9351766044123043,
  0.10703910470638403,
]
`;

exports[`shuffleArray shuffles list ["a", "b", "c", "d"] 1`] = `
Array [
  "c",
  "a",
  "d",
  "b",
]
`;

exports[`shuffleArray shuffles list ["a", "b", "c"] 1`] = `
Array [
  "b",
  "a",
  "c",
]
`;

exports[`shuffleArray shuffles list ["a", "b"] 1`] = `
Array [
  "b",
  "a",
]
`;

exports[`shuffleArray shuffles list ["a"] 1`] = `
Array [
  "a",
]
`;
